subprojects/process-services/src/main/java/org/gradle/process/internal/health/memory/DefaultOsMemoryInfo.java

```diff
@@ -16,22 +16,47 @@
 
 package org.gradle.process.internal.health.memory;
 
+import com.google.common.base.Objects;
 import org.gradle.internal.os.OperatingSystem;
+import com.google.common.io.Files;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.Charset;
 
 public class DefaultOsMemoryInfo implements OsMemoryInfo {
     private final OsMemoryInfo delegate;
+    private static final String CGROUP_MEM_USAGE_FILE = "/sys/fs/cgroup/memory/memory.usage_in_bytes";
+    private static final String CGROUP_MEM_TOTAL_FILE = "/sys/fs/cgroup/memory/memory.limit_in_bytes";
+    private static final String UNBOUNDED_MEMORY = "9223372036854771712";
 
     public DefaultOsMemoryInfo() {
         OperatingSystem operatingSystem = OperatingSystem.current();
         if (operatingSystem.isMacOsX()) {
             delegate = new NativeOsMemoryInfo();
         } else if (operatingSystem.isLinux()) {
-            delegate = new MemInfoOsMemoryInfo();
+            if (canUseCGroup()) {
+                delegate = new CGroupMemoryInfo();
+            } else {
+                delegate = new MemInfoOsMemoryInfo();
+            }
         } else {
             delegate = new MBeanOsMemoryInfo();
         }
     }
 
+    private boolean canUseCGroup() {
```

Comment:
Instead of having separate `MemoryInfo` objects and trying to determine whether cgroups are in play by checking against a magic value, maybe we should just have a single Linux `MemoryInfo` object that checks both meminfo and cgroups and then makes the decision about which value is best based on which one reports the lowest total physical memory.  Once you determine the correct source once, then you could just keep using that source for future updates.  Would something like that work?  Is there a situation where cgroups should not be used, but it reports a _lower_ value than meminfo?

Reply:
Yep that's a more sane approach, good idea. Will make this change promptly (sorry, I have only just now returned to this PR).

Reply:
Have made your proposed change